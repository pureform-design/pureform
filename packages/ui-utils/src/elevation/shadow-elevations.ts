import { deepMerge } from "@repo/utils";
import type { Pixel } from "../types";
import type {
    BaseElevation,
    BoxShadow,
    ElevationConfig,
    ShadowElevation,
} from "./types";
import {
    ambientMap,
    penumbraMap,
    precalculatedAtPixel,
    umbraMap,
} from "./precalculated";

type Config = ElevationConfig | Pixel;

export type ShadowElevationArgs<
    TCustomElevation extends string = BaseElevation,
> = { elevations: Partial<Record<BaseElevation | TCustomElevation, Config>> };

export class ShadowElevations<TCustomElevation extends string = BaseElevation> {
    private static defaultElevationOptions: ShadowElevationArgs = {
        elevations: {
            level0: 0,
            level1: 1,
            level2: 2,
            level3: 3,
            level4: 4,
            level5: 5,
        },
    };

    private elevationConfigs: Record<
        TCustomElevation | BaseElevation,
        ElevationConfig
    >;

    private elevationLevelCache: Record<string, ShadowElevation> = {};
    private elevationPxCache: Record<number, ShadowElevation> = {};

    public static create<TCustomElevation extends string = BaseElevation>(
        args?: ShadowElevationArgs<TCustomElevation>,
    ): ShadowElevations<TCustomElevation> {
        return new ShadowElevations<TCustomElevation>(args);
    }

    protected constructor(args?: ShadowElevationArgs<TCustomElevation>) {
        const options = deepMerge(
            ShadowElevations.defaultElevationOptions as ShadowElevationArgs<TCustomElevation>,
            args ?? {},
        );
        const configs = {} as Record<
            TCustomElevation | BaseElevation,
            ElevationConfig
        >;

        for (const key of Object.keys(options.elevations)) {
            const value = options.elevations[key as TCustomElevation];
            let numPx = -1;
            if (typeof value === "number" || typeof value === "string") {
                numPx =
                    typeof value === "number" ? value : Number.parseInt(value);
            } else if (typeof value === "object") {
                numPx =
                    typeof value.pixel === "number"
                        ? value.pixel
                        : Number.parseInt(value.pixel);
            } else {
                throw new Error(
                    "Invalid elevation value, must be a number or object with a pixel property",
                );
            }

            if (numPx < 0) {
                throw new Error(
                    "Invalid elevation pixel, must be greater than 0",
                );
            }

            configs[key as TCustomElevation] = {
                pixel: numPx,
            };
        }

        this.elevationConfigs = configs;
    }

    public get(elevation: TCustomElevation | BaseElevation): ShadowElevation {
        const cached = this.elevationLevelCache[elevation];

        if (cached) {
            return cached;
        }

        const config = this.elevationConfigs[elevation];

        if (!config) {
            throw new Error(`Invalid elevation: ${elevation}`);
        }

        const px = config.pixel;

        const numPx = typeof px === "number" ? px : Number.parseInt(px);

        const shadow = this.getAtPixel(numPx);

        this.elevationLevelCache[elevation] = shadow;

        return shadow;
    }

    // Penumbra functions
    private static penumbraOffsetY(px: number): number {
        const slope = 1.0;
        const intercept = 0.0;
        return Math.floor(slope * px + intercept);
    }

    private static penumbraBlurRadius(px: number): number {
        const slope = 1.37;
        const intercept = 1.18;
        return Math.floor(slope * px + intercept);
    }

    private static penumbraSpreadRadius(px: number): number {
        const slope = 0.13;
        const intercept = 0.03;
        return Math.floor(slope * px + intercept);
    }

    // Umbra functions
    private static umbraOffsetY(px: number): number {
        const slope = 0.39;
        const intercept = 0.63;
        return Math.floor(slope * px + intercept);
    }

    private static umbraBlurRadius(px: number): number {
        const slope = 0.56;
        const intercept = 1.02;
        return Math.floor(slope * px + intercept);
    }

    private static umbraSpreadRadius(px: number): number {
        const slope = -0.26;
        const intercept = -0.12;
        return Math.floor(slope * px + intercept);
    }

    // Ambient functions
    private static ambientOffsetY(px: number): number {
        const slope = 0.39;
        const intercept = 0.63;
        return Math.floor(slope * px + intercept);
    }

    private static ambientBlurRadius(px: number): number {
        const slope = 1.52;
        const intercept = 2.33;
        return Math.floor(slope * px + intercept);
    }

    private static ambientSpreadRadius(px: number): number {
        const slope = 0.29;
        const intercept = 0.0;
        return Math.floor(slope * px + intercept);
    }

    private static umbraAtPixel(px: number): BoxShadow {
        return {
            offsetX: "0px",
            offsetY: `${ShadowElevations.umbraOffsetY(px)}px`,
            blurRadius: `${ShadowElevations.umbraBlurRadius(px)}px`,
            spreadRadius: `${ShadowElevations.umbraSpreadRadius(px)}px`,
        };
    }

    private static penumbraAtPixel(px: number): BoxShadow {
        return {
            offsetX: "0px",
            blurRadius: `${ShadowElevations.penumbraBlurRadius(px)}px`,
            offsetY: `${ShadowElevations.penumbraOffsetY(px)}px`,
            spreadRadius: `${ShadowElevations.penumbraSpreadRadius(px)}px`,
        };
    }

    private static ambientAtPixel(px: number): BoxShadow {
        return {
            offsetX: "0px",
            blurRadius: `${ShadowElevations.ambientBlurRadius(px)}px`,
            offsetY: `${ShadowElevations.ambientOffsetY(px)}px`,
            spreadRadius: `${ShadowElevations.ambientSpreadRadius(px)}px`,
        };
    }

    public getAtPixel(px: number): ShadowElevation {
        if (px < 0) {
            throw new Error("Invalid elevation pixel, must be greater than 0");
        }

        const cached = this.elevationPxCache[px];

        if (cached) {
            return cached;
        }

        if (px >= 0 && px <= 23) {
            return precalculatedAtPixel(px);
        }

        const shadow: ShadowElevation = {
            umbra: ShadowElevations.umbraAtPixel(px),
            penumbra: ShadowElevations.penumbraAtPixel(px),
            ambient: ShadowElevations.ambientAtPixel(px),
        };

        this.elevationPxCache[px] = shadow;

        return shadow;
    }
}
