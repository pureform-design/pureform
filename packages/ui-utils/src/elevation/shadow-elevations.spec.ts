import { describe, expect, it } from "vitest";
import { ShadowElevations } from "./shadow-elevations";
import type { BaseElevation } from "./types";

describe("ShadowElevations", () => {
    it("should be able to create a shadow elevation instance", () => {
        const se = ShadowElevations.create();

        expect(se).toBeInstanceOf(ShadowElevations);
    });

    it("should be able to get a shadow elevation with a level of 0", () => {
        const se = ShadowElevations.create();

        const shadow = se.get("level0");

        expect(shadow).toEqual({
            umbra: {
                offsetX: "0px",
                offsetY: "0px",
                blurRadius: "0px",
                spreadRadius: "0px",
            },
            penumbra: {
                offsetX: "0px",
                offsetY: "0px",
                blurRadius: "0px",
                spreadRadius: "0px",
            },
            ambient: {
                offsetX: "0px",
                offsetY: "0px",
                blurRadius: "0px",
                spreadRadius: "0px",
            },
        });
    });

    it("should be able to get a shadow elevation with a level of 1 to 5", () => {
        const se = ShadowElevations.create();

        for (let i = 1; i <= 5; i++) {
            const shadow = se.get(`level${i}` as BaseElevation);

            expect(shadow).toBeTypeOf("object");
        }
    });

    it.for(new Array(24).fill(0).map((_, i) => i + 1))(
        "should match the snapshot shadow elevation for %i",
        (i) => {
            const se = ShadowElevations.create();

            const shadow = se.getAtPixel(i);
            expect(shadow).toMatchSnapshot();
        },
    );

    it("should be able to get a shadow elevation using custom elevation", () => {
        const se = ShadowElevations.create({
            elevations: {
                level10: 10,
                level20: 20,
            },
        });

        const shadow = se.get("level10");
        const shadow2 = se.get("level20");

        expect(shadow).toBeTypeOf("object");
        expect(shadow2).toBeTypeOf("object");
    });

    it("should throw when trying to get a shadow elevation with an invalid elevation", () => {
        const se = ShadowElevations.create();

        expect(() => se.get("invalid" as BaseElevation)).toThrow();
    });

    it("should throw when trying to get a shadow elevation with an negative pixel", () => {
        const se = ShadowElevations.create();

        expect(() => se.getAtPixel(-1)).toThrow();
    });
});
