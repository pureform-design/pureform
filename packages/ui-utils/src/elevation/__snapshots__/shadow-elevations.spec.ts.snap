// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShadowElevations > should match the snapshot shadow elevation for 1 1`] = `
{
  "ambient": {
    "blurRadius": "3px",
    "offsetX": "0px",
    "offsetY": "1px",
    "spreadRadius": "0px",
  },
  "penumbra": {
    "blurRadius": "1px",
    "offsetX": "0px",
    "offsetY": "1px",
    "spreadRadius": "0px",
  },
  "umbra": {
    "blurRadius": "1px",
    "offsetX": "0px",
    "offsetY": "2px",
    "spreadRadius": "-1px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 2 1`] = `
{
  "ambient": {
    "blurRadius": "5px",
    "offsetX": "0px",
    "offsetY": "1px",
    "spreadRadius": "0px",
  },
  "penumbra": {
    "blurRadius": "2px",
    "offsetX": "0px",
    "offsetY": "2px",
    "spreadRadius": "0px",
  },
  "umbra": {
    "blurRadius": "1px",
    "offsetX": "0px",
    "offsetY": "3px",
    "spreadRadius": "-2px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 3 1`] = `
{
  "ambient": {
    "blurRadius": "8px",
    "offsetX": "0px",
    "offsetY": "1px",
    "spreadRadius": "0px",
  },
  "penumbra": {
    "blurRadius": "4px",
    "offsetX": "0px",
    "offsetY": "3px",
    "spreadRadius": "0px",
  },
  "umbra": {
    "blurRadius": "3px",
    "offsetX": "0px",
    "offsetY": "3px",
    "spreadRadius": "-2px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 4 1`] = `
{
  "ambient": {
    "blurRadius": "10px",
    "offsetX": "0px",
    "offsetY": "1px",
    "spreadRadius": "0px",
  },
  "penumbra": {
    "blurRadius": "5px",
    "offsetX": "0px",
    "offsetY": "4px",
    "spreadRadius": "0px",
  },
  "umbra": {
    "blurRadius": "4px",
    "offsetX": "0px",
    "offsetY": "2px",
    "spreadRadius": "-1px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 5 1`] = `
{
  "ambient": {
    "blurRadius": "14px",
    "offsetX": "0px",
    "offsetY": "1px",
    "spreadRadius": "0px",
  },
  "penumbra": {
    "blurRadius": "8px",
    "offsetX": "0px",
    "offsetY": "5px",
    "spreadRadius": "0px",
  },
  "umbra": {
    "blurRadius": "5px",
    "offsetX": "0px",
    "offsetY": "3px",
    "spreadRadius": "-1px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 6 1`] = `
{
  "ambient": {
    "blurRadius": "18px",
    "offsetX": "0px",
    "offsetY": "1px",
    "spreadRadius": "0px",
  },
  "penumbra": {
    "blurRadius": "10px",
    "offsetX": "0px",
    "offsetY": "6px",
    "spreadRadius": "0px",
  },
  "umbra": {
    "blurRadius": "5px",
    "offsetX": "0px",
    "offsetY": "3px",
    "spreadRadius": "-1px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 7 1`] = `
{
  "ambient": {
    "blurRadius": "16px",
    "offsetX": "0px",
    "offsetY": "2px",
    "spreadRadius": "1px",
  },
  "penumbra": {
    "blurRadius": "10px",
    "offsetX": "0px",
    "offsetY": "7px",
    "spreadRadius": "1px",
  },
  "umbra": {
    "blurRadius": "5px",
    "offsetX": "0px",
    "offsetY": "4px",
    "spreadRadius": "-2px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 8 1`] = `
{
  "ambient": {
    "blurRadius": "14px",
    "offsetX": "0px",
    "offsetY": "3px",
    "spreadRadius": "2px",
  },
  "penumbra": {
    "blurRadius": "10px",
    "offsetX": "0px",
    "offsetY": "8px",
    "spreadRadius": "1px",
  },
  "umbra": {
    "blurRadius": "5px",
    "offsetX": "0px",
    "offsetY": "5px",
    "spreadRadius": "-3px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 9 1`] = `
{
  "ambient": {
    "blurRadius": "16px",
    "offsetX": "0px",
    "offsetY": "3px",
    "spreadRadius": "2px",
  },
  "penumbra": {
    "blurRadius": "12px",
    "offsetX": "0px",
    "offsetY": "9px",
    "spreadRadius": "1px",
  },
  "umbra": {
    "blurRadius": "6px",
    "offsetX": "0px",
    "offsetY": "5px",
    "spreadRadius": "-3px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 10 1`] = `
{
  "ambient": {
    "blurRadius": "18px",
    "offsetX": "0px",
    "offsetY": "4px",
    "spreadRadius": "3px",
  },
  "penumbra": {
    "blurRadius": "14px",
    "offsetX": "0px",
    "offsetY": "10px",
    "spreadRadius": "1px",
  },
  "umbra": {
    "blurRadius": "6px",
    "offsetX": "0px",
    "offsetY": "6px",
    "spreadRadius": "-3px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 11 1`] = `
{
  "ambient": {
    "blurRadius": "20px",
    "offsetX": "0px",
    "offsetY": "4px",
    "spreadRadius": "3px",
  },
  "penumbra": {
    "blurRadius": "15px",
    "offsetX": "0px",
    "offsetY": "11px",
    "spreadRadius": "1px",
  },
  "umbra": {
    "blurRadius": "7px",
    "offsetX": "0px",
    "offsetY": "6px",
    "spreadRadius": "-4px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 12 1`] = `
{
  "ambient": {
    "blurRadius": "22px",
    "offsetX": "0px",
    "offsetY": "5px",
    "spreadRadius": "4px",
  },
  "penumbra": {
    "blurRadius": "17px",
    "offsetX": "0px",
    "offsetY": "12px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "8px",
    "offsetX": "0px",
    "offsetY": "7px",
    "spreadRadius": "-4px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 13 1`] = `
{
  "ambient": {
    "blurRadius": "24px",
    "offsetX": "0px",
    "offsetY": "5px",
    "spreadRadius": "4px",
  },
  "penumbra": {
    "blurRadius": "19px",
    "offsetX": "0px",
    "offsetY": "13px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "8px",
    "offsetX": "0px",
    "offsetY": "7px",
    "spreadRadius": "-4px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 14 1`] = `
{
  "ambient": {
    "blurRadius": "26px",
    "offsetX": "0px",
    "offsetY": "5px",
    "spreadRadius": "4px",
  },
  "penumbra": {
    "blurRadius": "21px",
    "offsetX": "0px",
    "offsetY": "14px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "9px",
    "offsetX": "0px",
    "offsetY": "7px",
    "spreadRadius": "-4px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 15 1`] = `
{
  "ambient": {
    "blurRadius": "28px",
    "offsetX": "0px",
    "offsetY": "6px",
    "spreadRadius": "5px",
  },
  "penumbra": {
    "blurRadius": "22px",
    "offsetX": "0px",
    "offsetY": "15px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "9px",
    "offsetX": "0px",
    "offsetY": "8px",
    "spreadRadius": "-5px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 16 1`] = `
{
  "ambient": {
    "blurRadius": "30px",
    "offsetX": "0px",
    "offsetY": "6px",
    "spreadRadius": "5px",
  },
  "penumbra": {
    "blurRadius": "24px",
    "offsetX": "0px",
    "offsetY": "16px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "10px",
    "offsetX": "0px",
    "offsetY": "8px",
    "spreadRadius": "-5px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 17 1`] = `
{
  "ambient": {
    "blurRadius": "32px",
    "offsetX": "0px",
    "offsetY": "6px",
    "spreadRadius": "5px",
  },
  "penumbra": {
    "blurRadius": "26px",
    "offsetX": "0px",
    "offsetY": "17px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "11px",
    "offsetX": "0px",
    "offsetY": "8px",
    "spreadRadius": "-5px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 18 1`] = `
{
  "ambient": {
    "blurRadius": "34px",
    "offsetX": "0px",
    "offsetY": "7px",
    "spreadRadius": "6px",
  },
  "penumbra": {
    "blurRadius": "28px",
    "offsetX": "0px",
    "offsetY": "18px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "11px",
    "offsetX": "0px",
    "offsetY": "9px",
    "spreadRadius": "-5px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 19 1`] = `
{
  "ambient": {
    "blurRadius": "36px",
    "offsetX": "0px",
    "offsetY": "7px",
    "spreadRadius": "6px",
  },
  "penumbra": {
    "blurRadius": "29px",
    "offsetX": "0px",
    "offsetY": "19px",
    "spreadRadius": "2px",
  },
  "umbra": {
    "blurRadius": "12px",
    "offsetX": "0px",
    "offsetY": "9px",
    "spreadRadius": "-6px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 20 1`] = `
{
  "ambient": {
    "blurRadius": "38px",
    "offsetX": "0px",
    "offsetY": "8px",
    "spreadRadius": "7px",
  },
  "penumbra": {
    "blurRadius": "31px",
    "offsetX": "0px",
    "offsetY": "20px",
    "spreadRadius": "3px",
  },
  "umbra": {
    "blurRadius": "13px",
    "offsetX": "0px",
    "offsetY": "10px",
    "spreadRadius": "-6px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 21 1`] = `
{
  "ambient": {
    "blurRadius": "40px",
    "offsetX": "0px",
    "offsetY": "8px",
    "spreadRadius": "7px",
  },
  "penumbra": {
    "blurRadius": "33px",
    "offsetX": "0px",
    "offsetY": "21px",
    "spreadRadius": "3px",
  },
  "umbra": {
    "blurRadius": "13px",
    "offsetX": "0px",
    "offsetY": "10px",
    "spreadRadius": "-6px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 22 1`] = `
{
  "ambient": {
    "blurRadius": "42px",
    "offsetX": "0px",
    "offsetY": "8px",
    "spreadRadius": "7px",
  },
  "penumbra": {
    "blurRadius": "35px",
    "offsetX": "0px",
    "offsetY": "22px",
    "spreadRadius": "3px",
  },
  "umbra": {
    "blurRadius": "14px",
    "offsetX": "0px",
    "offsetY": "10px",
    "spreadRadius": "-6px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 23 1`] = `
{
  "ambient": {
    "blurRadius": "44px",
    "offsetX": "0px",
    "offsetY": "9px",
    "spreadRadius": "8px",
  },
  "penumbra": {
    "blurRadius": "36px",
    "offsetX": "0px",
    "offsetY": "23px",
    "spreadRadius": "3px",
  },
  "umbra": {
    "blurRadius": "14px",
    "offsetX": "0px",
    "offsetY": "11px",
    "spreadRadius": "-7px",
  },
}
`;

exports[`ShadowElevations > should match the snapshot shadow elevation for 24 1`] = `
{
  "ambient": {
    "blurRadius": "38px",
    "offsetX": "0px",
    "offsetY": "9px",
    "spreadRadius": "6px",
  },
  "penumbra": {
    "blurRadius": "34px",
    "offsetX": "0px",
    "offsetY": "24px",
    "spreadRadius": "3px",
  },
  "umbra": {
    "blurRadius": "14px",
    "offsetX": "0px",
    "offsetY": "9px",
    "spreadRadius": "-7px",
  },
}
`;
